### ISSUE TITLE
The runtime can't apply program-defined credit operations concurrently

### ISSUE THREAD

**Issue by garious** [solana-labs/solana#1470]
When transferring native tokens (sols), the runtime will allow you to credit the same account *concurrently* from two transactions within the same Entry. The runtime handles this case explicitly.

Problem for program-defined tokens:

Parallel credits to the same account is not available to program-defined tokens, such as erc20 tokens.  That required program_id also indicates that only one contract may operate on that account within that ledger Entry.

Proposed solution:

The program could define functions which tell the runtime what a safe change to its tokens and userdata looks like.  For the System program, it'd be the change in tokens.  In an erc20 token, for example, it'd be the change in an integer in the account's userdata.

```rust
type Delta = Vec<u8>;
fn get_delta(old: &Account, new: &Account) -> Result<Delta>;
fn apply_delta(old: &mut Account, delta &Delta);
```

For SystemProgram, that'd be:

```rust
fn get_delta(old: &Account, new: &Account) -> Result<Delta> {
   if new.tokens < old.tokens {
       return Err(InvalidCredit);
   }
   Ok(serialize(&(new.tokens - old.tokens)))
}

fn apply_delta(old: &mut Account, delta &Delta) {
   old.tokens += deserialize(delta);
}
```

**Comment by garious** [comment]
@jackcmay, @mvines, @aeyakovenko, @CriesofCarrots fyi

**Comment by garious** [comment]
@aeyakovenko, a `signed` flag would solve the same problem that the old `pending` vector solved - running contracts that were authorized to spend a fixed amount of tokens. Adding a `signed` flag is a cleaner solution though, because it leaves the account in `accounts`, allowing users to credit it with additional tokens.

**Comment by aeyakovenko** [comment]
@garious you can do a batched spend or credit to the same account with a single TX.  what kind of transaction are you trying to build?

**Comment by garious** [comment]
`get_delta`/`apply_delta` is a lousy design. It'd allow programs to inject nondeterministic behavior.  Instead, the solana-runtime would need to define a set of primitive operations (anything in the category Group). After that we'd need a mechanism to add type annotations to `Account.data`.   If we did all that, the runtime could guarantee that only "credit" operations were performed on the account.

tldr; this issue needs a design proposal to move forward
